UIPS/UIP-0122.md

```diff
@@ -0,0 +1,190 @@
+---
+uip: "0122"
+title: "Lagoon IEEE 754 Reals"
+description: Support jetted array operations
+author: ~lagrev-nocfep
+status: Draft
+type: Standards Track
+category: Arvo
+created: 2024-03-26
+---
+
+## Abstract
+
+Nock and its predicated software programs are designed to facilitate computing as a legible state machine. At-scale scientific computing has never really been part of the design criteria for the Urbit ecosystem. While it's hard to imagine density functional theory quantum chemistry packages preferring Hoon, it is far easier to envision a world of machine learning and even LLMs that are built on top of or adjacent to Urbit. At their core, essentially all of the “scientific computing” approaches rely heavily on linear algebra packages: classically BLAS and LAPACK, but many others as well today. It behooves us to provide this to future developers building on Mars.
+
+Lagoon has matured through several refactors into a developer-friendly interface.  It is time to prepare and release a subset of completed array-operational work to the developer community.  Since Lagoon relies on jets to operate reasonably quickly, we will prepare and release `%real`-valued operations first.
+
+## Motivation
+
+Lagoon (Linear AlGebra in hOON) will offer BLAS-like vector and matrix operations (like NumPy's linear algebra essentials, but not solvers).  This will facilitate a common and performant interface for all developers relying on a common lagnuage of data types and operators to represent vector, matrix, and tensor data.  By focusing on `%real`s first, we can vet the interface further in production and lay the groundwork for expansion in further types in subsequent releases.
+
+## Specification
+
+### Data Types
+
+Lagoon defines the following data types for `%real`-valued numbers (which are IEEE 754 floats compatible with corresponding `@r` types in Hoon).  (Note that we are commenting out other prospective types for which some work has been done.)
+
+```hoon
+  ::  /sur/lagoon
+::::  Types for Lagoon compatibility
+::
+|%
++$  ray               ::  $ray:  n-dimensional array
+  $:  =meta           ::  descriptor
+      data=@ux        ::  data, row-major order
+  ==
+::
++$  prec  [a=@ b=@]   ::  fixed-point precision, a+b+1=bloq
++$  meta              ::  $meta:  metadata for a $ray
+  $:  shape=(list @)  ::  list of dimension lengths
+      =bloq           ::  logarithm of bitwidth
+      =kind           ::  name of data type
+      fxp=(unit prec) ::  fixed-point scale
```

Comment:
do we want to comment out fxp?

Reply:
How finicky are state upgrades going to be with adding a value here, I guess is the question.

Reply:
The real PITA is that we will have to `s///g` all the test cases.